## 1. RNA-Seq Expression Analysis

### 1.1. Introduction

RNA sequencing (RNA-Seq) is a high-throughput method used to profile the transcriptome, quan- tify gene expression and discover novel RNA molecules or splice variants. This tutorial uses RNA sequencing of human cancer samples to walk you through transcriptome alignment, visualisation, simple quality control checks and shows you how to profile transcriptomic differences by identifying differentially expressed genes.

For an introduction to RNA-Seq principles and best practices see:

> **A survey of best practices for RNA-Seq data analysis**  
Ana Conesa, Pedro Madrigal, Sonia Tarazona, David Gomez-Cabrero, Alejandra Cervera, Andrew McPherson, Michał Wojciech Szcześniak, Daniel J. Gaffney, Laura L. Elo, Xue- gong Zhang and Ali Mortazavi Genome Biol. 2016 Jan 26;17:13 doi:10.1186/s13059-016-0881-8

### 1.2. Learning Outcomes

By the end of this tutorial you can expect to be able to:
<ul>
	<li> Align RNA-Seq reads to a reference genome and a transcriptome </li>
	<li> Visualise transcription data using standard tools </li>
	<li> Perform QC of NGS transcriptomic data </li>
	<li> Quantify the expression values of your transcripts using standard tools </li>
</ul>

### 1.3. Practical Outline
This tutorial comprises the following sections:  
<ol>
	<li>Introducing the tutorial dataset  
	<li>Mapping RNA-Seq reads to the genome with HISAT2  
	<li>Visualising transcriptomes with IGV  
	<li>Transcript quantification with Kallisto  
	<li>Identifying differentially expressed genes with Sleuth  
	<li>Key aspects of differential expression analysis
</ol>

## 2. Introducing The Tutorial Dataset

## 3. Mapping RNA-Seq Reads to the Genome using `HISAT2`

## 4. Visualising transcriptomes with IGV

## 5. Transcript quantification with Kallisto

## 6. Differential Expression Analysis With Sleuth

## 7. Key Aspects of Differential Expression Analysis

## 8. Normalisation



